File: 2asir/web-deployment/PHP/clase/OBJETOS/PHP_POO_Mastery_Kit/02_Estructuras_Control.php
<?php
/**
 * 02_Estructuras_Control.php
 * 
 * OBJETIVO:
 * Controlar el ciclo de vida de un objeto y aplicar lógica dentro de las clases.
 * 
 * CONCEPTOS:
 * 1. Constructor (__construct): Método mágico que se ejecuta AUTOMÁTICAMENTE al hacer `new`.
 * 2. Destructor (__destruct): Se ejecuta cuando el objeto se elimina o acaba el script.
 * 3. Lógica interna: Usar if/else dentro de los métodos para validar datos.
 */

class CuentaBancaria
{
    public $titular;
    public $saldo;

    // El CONSTRUCTOR inicializa el objeto con datos obligatorios
    public function __construct($nombreTitular, $saldoInicial = 0)
    {
        $this->titular = $nombreTitular;
        $this->saldo = $saldoInicial;
        echo "[INFO] Cuenta creada para {$this->titular} con {$this->saldo}€.\n";
    }

    public function depositar($cantidad)
    {
        // Estructura de control para validar
        if ($cantidad > 0) {
            $this->saldo += $cantidad;
            echo "Se han depositado {$cantidad}€. Nuevo saldo: {$this->saldo}€.\n";
        } else {
            echo "[ERROR] La cantidad a depositar debe ser positiva.\n";
        }
    }

    public function retirar($cantidad)
    {
        // Lógica para evitar saldo negativo
        if ($cantidad > $this->saldo) {
            echo "[ERROR] Fondos insuficientes. Tienes {$this->saldo}€ y quieres sacar {$cantidad}€.\n";
        } elseif ($cantidad <= 0) {
            echo "[ERROR] Debes retirar una cantidad válida.\n";
        } else {
            $this->saldo -= $cantidad;
            echo "Se han retirado {$cantidad}€. Nuevo saldo: {$this->saldo}€.\n";
        }
    }

    // El DESTRUCTOR se llama al final útiles para cerrar conexiones o guardar logs
    public function __destruct()
    {
        echo "[CIERRE] Cerrando sesión bancaria de {$this->titular}.\n";
    }
}

## -----------------------------------------
## SECCIÓN DE PRUEBAS
## -----------------------------------------

echo "--- INICIO DEL PROGRAMA ---\n";

// Al hacer new, se ejecuta __construct solo
$miCuenta = new CuentaBancaria("Juan Pérez", 100);

// Probamos la lógica de control
$miCuenta->depositar(50);   // Funciona
$miCuenta->depositar(-20);  // Error validado
$miCuenta->retirar(200);    // Error fondos insuficientes
$miCuenta->retirar(30);     // Funciona

echo "--- FIN DEL PROGRAMA (El destructor se ejecutará ahora) ---\n";

/*
 * NOTA DOCENTE:
 * El constructor es vital para asegurar que un objeto nunca exista "vacío" o en un estado inválido.
 * Obliga al programador a dar los datos mínimos necesarios al crear el objeto.
 */
?>

File: 2asir/web-deployment/PHP/clase/OBJETOS/Python_POO_Mastery_Kit/02_Estructuras_Control.py
"""
02_Estructuras_Control.py

OBJETIVO:
Controlar el ciclo de vida y lógica interna.

CONCEPTOS CLAVE vs PHP:
- Constructor: `__construct` (PHP) -> `__init__` (Python)
- Destructor: `__destruct` (PHP) -> `__del__` (Python)
- Concatenación: `.` (PHP) -> `+` o f-strings (Python)
"""

class CuentaBancaria:
    
    # El CONSTRUCTOR (__init__)
    # 'self' es obligatorio ponerlo, pero no enviarlo al llamar.
    def __init__(self, nombre_titular, saldo_inicial=0):
        self.titular = nombre_titular
        self.saldo = saldo_inicial
        print(f"[INFO] Cuenta creada para {self.titular} con {self.saldo}€.")
    
    def depositar(self, cantidad):
        if cantidad > 0:
            self.saldo += cantidad
            print(f"Se han depositado {cantidad}€. Nuevo saldo: {self.saldo}€.")
        else:
            print("[ERROR] La cantidad a depositar debe ser positiva.")
    
    def retirar(self, cantidad):
        if cantidad > self.saldo:
            print(f"[ERROR] Fondos insuficientes. Tienes {self.saldo}€ y quieres sacar {cantidad}€.")
        elif cantidad <= 0:
            print("[ERROR] Debes retirar una cantidad válida.")
        else:
            self.saldo -= cantidad
            print(f"Se han retirado {cantidad}€. Nuevo saldo: {self.saldo}€.")
            
    # El DESTRUCTOR (__del__)
    # En Python es raro usarlo manualmente, el Garbage Collector se encarga.
    def __del__(self):
        print(f"[CIERRE] Cerrando sesión bancaria de {self.titular}.")

## -----------------------------------------
## SECCIÓN DE PRUEBAS
## -----------------------------------------

print("--- INICIO DEL PROGRAMA ---")

mi_cuenta = CuentaBancaria("Juan Pérez", 100)

mi_cuenta.depositar(50)
mi_cuenta.depositar(-20)
mi_cuenta.retirar(200)
mi_cuenta.retirar(30)

print("--- FIN DEL PROGRAMA ---")
# El destructor se llamará automáticamente cuando 'mi_cuenta' deje de ser útil
